src/resource_definition.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""gateway-api-integrator resource definition."""
+
+import dataclasses
+import itertools
+import typing
+
+import ops
+from pydantic import BaseModel, Field, ValidationError
+
+
+class InvalidCharmConfigError(Exception):
+    """Exception raised when a charm configuration is found to be invalid.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the InvalidCharmConfigError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class CharmConfig(BaseModel):
+    """Charm configuration.
+
+    Attrs:
+        gateway_class (_type_): _description_
+        external_hostname: The configured gateway hostname.
+    """
+
+    gateway_class: str = Field(min_length=1)
+    external_hostname: str = Field(
+        pattern=r"[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
```

Comment:
This is the same as the one used by NGINX charm, right? I wonder if wouldn't be nice to have some kind of library with all the regex that we are using

Reply:
Yep, a lib implementation is in the backlog i think, but it'll be way down the line